src/larecs/comptime_optional.mojo

```diff
@@ -0,0 +1,133 @@
+from memory import UnsafePointer
+
+
+@value
+struct ComptimeOptional[
+    ElementType: CollectionElement,
+    has_value: Bool = True,
+](Movable, Copyable, ExplicitlyCopyable):
+    """An optional type that can potentially hold a value of ElementType.
+
+    In contrast to the built-in oprtional, it is decided at
+    compile time whether a value is present. This allows
+    this type to have a size of 0 in these cases.
+
+    Parameters:
+        ElementType: The type of the elements in the array.
+        has_value: Whether the element exists.
+    """
+
+    # Fields
+    alias type = __mlir_type[
+        `!pop.array<`, Int(has_value).value, `, `, Self.ElementType, `>`
+    ]
+    var _value: Self.type
+    """The underlying storage for the optional."""
+
+    # ===------------------------------------------------------------------===#
+    # Life cycle methods
+    # ===------------------------------------------------------------------===#
+
+    @always_inline
+    @implicit
+    fn __init__(out self, none: None = None):
+        """This constructor will always cause a compile time error if used.
+        It is used to steer users away from uninitialized memory.
+        """
+        constrained[
+            not has_value,
+            "Initialize with a value if `has_value` is `True`",
+        ]()
+        __mlir_op.`lit.ownership.mark_initialized`(__get_mvalue_as_litref(self))
+
+    @always_inline
+    @implicit
+    fn __init__(out self, owned value: Self.ElementType):
+        """Constructs an optional type holding the provided value.
+
+        Args:
+            value: The value to fill the optional with.
+        """
+        __mlir_op.`lit.ownership.mark_initialized`(__get_mvalue_as_litref(self))
+
+        var ptr = UnsafePointer.address_of(self.value())
```

Comment:
Does this really need to use unsafe? Can't you just store the value in `_value`?